src/job_shop_scheduler_cqm.py

```diff
@@ -0,0 +1,237 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import dimod
+
+from dwave.system import LeapHybridCQMSampler
+
+from src.model_data import JobShopData
+
+
+def sum_to_one(*args):
+    return sum(args) == 1
+
+
+def get_label(task, time):
+    """Creates a standardized name for variables in the constraint satisfaction problem,
+    JobShopScheduler.csp.
+    """
+    return "{task.job}_{task.resource},{time}".format(**locals())
+
+
+class JobShopScheduler:
+    def __init__(self, model_data: JobShopData):
+        """
+        Args:
+            model_data: A JobShopData object that holds the data for this job shop 
+            scheduling problem.
+
+        """
+        self.model_data = model_data
+
+        self._process_data()
+        self._initialize_variables()
+        self.cqm = dimod.ConstrainedQuadraticModel()
+        self._add_task_start_constraint()
+        self._add_precedence_constraint()
+        self._add_share_machine_constraint()
+        self._remove_absurd_times()
+        # self._model_end()
+        self._add_objective()
+
+
+    def solve(self, profile: str=None) -> dict:
+        """Solves the job shop scheduling problem.
+        
+        Args:
+            profile: A string. The name of the profile to use for the hybrid solver.
+
+        Returns:
+            dict: A dictionary of the form {task: (start_time, end_time)} for each task in the
+            problem.
+        """
+        sampler = LeapHybridCQMSampler(profile=profile)
+        res = sampler.sample_cqm(self.cqm, time_limit=10, label='Job Shop Demo')
+
+        res.resolve()
+        feasible_sampleset = res.filter(lambda d: d.is_feasible)
+        try:
+            best_feasible = feasible_sampleset.first.sample
+
+        except ValueError:
+            raise RuntimeError(
+                "Sampleset is empty, try increasing time limit or " +
+                "adjusting problem config."
+            )
+
+        task_assignments = {}
+        for task in self.model_data.get_tasks():
+            for t in range(self.max_time):
+                if best_feasible[get_label(task, t)] == 1:
+                    start_time = t
+                    end_time = t + task.duration
+                    task_assignments[task] = (start_time, end_time)
+                    break
+
+        return task_assignments
+
+    
+    def _initialize_variables(self) -> None:
+        """Initialize the variables for the constraint satisfaction problem.
+        """
+        task_start_vars = {get_label(task, t): dimod.Binary(get_label(task, t)) 
```

Comment:
If model construction time is slow, we could do it using labels rather than symbols. See [Scaling for Production](https://docs.ocean.dwavesys.com/en/stable/docs_dimod/intro/intro_scaling.html). But if it's not an issue then I think this is fine.